connectors/sources/graphql.py

```diff
@@ -0,0 +1,449 @@
+        if re.match(URL_regex, self.graphql_client.url):
+            return True
+        return False
+
+    async def validate_config(self):
+        """Validates whether user input is empty or not for configuration fields
+        Also validate, if user configured repositories are accessible or not and scope of the token
+        """
+        await super().validate_config()
+
+        if not self.validate_endpoints():
+            msg = "HTTP URL & Endpoint are not structured."
+            raise ConfigurableFieldValueError(msg)
+
+        if not self.is_query(graphql_query=self.graphql_client.graphql_query):
+            msg = "Configured Query is not supported by the connector."
+            raise ConfigurableFieldValueError(msg)
+
+        headers = self.graphql_client.configuration["headers"]
+        graphql_variables = self.graphql_client.configuration["graphql_variables"]
+        if headers:
+            try:
+                self.graphql_client.headers = json.loads(headers)
+            except Exception as exception:
+                msg = f"Error while processing configured GraphQL headers. Exception: {exception}"
+                raise ConfigurableFieldValueError(msg) from exception
+
+        if graphql_variables:
+            try:
+                self.graphql_client.variables = json.loads(graphql_variables)
+            except Exception as exception:
+                msg = f"Error while processing configured GraphQL variables. Exception: {exception}"
+                raise ConfigurableFieldValueError(msg) from exception
+
+    async def close(self):
+        await self.graphql_client.close()
+
+    async def ping(self):
+        try:
+            await self.graphql_client.ping()
+            self._logger.debug("Successfully connected to GraphQL Instance.")
+        except Exception:
+            self._logger.exception("Error while connecting to GraphQL Instance.")
+            raise
+
+    async def fetch_data(self, graphql_query):
+        if self.graphql_client.pagination_enabled:
+            async for data in self.graphql_client.paginated_call(
+                graphql_query=graphql_query
+            ):
+                if isinstance(data, dict):
+                    yield data
+                elif isinstance(data, list):
+                    for document in data:  # pyright: ignore
+                        if isinstance(document, dict):
+                            yield document
+        else:
+            data = await self.graphql_client.make_request(graphql_query=graphql_query)
+            for documents in self.graphql_client.extract_graphql_data_items(
+                key="data", data=data
+            ):
+                if isinstance(documents, dict):
+                    yield documents
+                elif isinstance(documents, list):
+                    for document in documents:
+                        if isinstance(document, dict):
+                            yield document
```

Comment:
This is duplicated code, we have same logic in 414 to 419

Reply:
add this piece of code into a function